### PR TITLE
Profile: Fix tests when run with nthreads > 1 (for buildkite)

### LINKED ISSUE #42252: `Profile` tests fail when multithreading is enabled
## Minimum working example

Tests pass when multithreading is disabled:

```
$ export JULIA_NUM_THREADS=1
$ ./julia -e 'Base.runtests(["Profile"]; ncores = 16)'
Test  (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
Profile    (1) |        started at 2021-09-15T03:16:53.811
Profile    (1) |    34.66 |   1.44 |  4.1 |     757.84 |   350.34

Test Summary: | Pass  Total
  Overall     |   77     77
    SUCCESS
```

Tests fail when multithreading is enabled:

```
$ export JULIA_NUM_THREADS=16
$ ./julia -e 'Base.runtests(["Profile"]; ncores = 16)'
Test  (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
Profile    (1) |        started at 2021-09-15T03:21:14.877
Profile    (1) |         failed at 2021-09-15T03:21:33.589
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:110
  Expression: n_ == def_n
   Evaluated: 160000000 == 10000000
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:115
  Expression: n_ == 1000001
   Evaluated: 16000016 == 1000001
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:123
  Expression: Profile.fetch()
   Evaluated: Test.LogRecord[]


Test Summary: | Pass  Fail  Total
  Overall     |   74     3     77
    Profile   |   74     3     77
    FAILURE

The global RNG seed was 0x818b54fd081378cf6abebfde2a083c4f.

Error in testset Profile:
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:110
  Expression: n_ == def_n
   Evaluated: 160000000 == 10000000
Error in testset Profile:
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:115
  Expression: n_ == 1000001
   Evaluated: 16000016 == 1000001
Error in testset Profile:
Test Failed at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/stdlib/v1.8/Profile/test/runtests.jl:123
  Expression: Profile.fetch()
   Evaluated: Test.LogRecord[]
ERROR: LoadError: Test run finished with errors
in expression starting at /Users/dilum/dev/repos/JuliaLang/julia/usr/share/julia/test/runtests.jl:93
ERROR: A test has failed. Please submit a bug report (https://github.com/JuliaLang/julia/issues)
including error messages above and the output of versioninfo():
Julia Version 1.8.0-DEV.542
Commit 3794f9a307 (2021-09-15 06:04 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin20.6.0)
  CPU: Intel(R) Core(TM) i5-4278U CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, haswell)
Environment:
  JULIA_NUM_THREADS = 16
  JULIA_PKG_SERVER =
  JULIA_CUDA_VERBOSE = true

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] runtests(tests::Vector{String}; ncores::Int64, exit_on_error::Bool, revise::Bool, seed::Nothing)
   @ Base ./util.jl:576
 [3] top-level scope
   @ none:1

caused by: failed process: Process(setenv(`/Users/dilum/dev/repos/JuliaLang/julia/usr/bin/julia -Cnative -J/Users/dilum/dev/repos/JuliaLang/julia/usr/lib/julia/sys.dylib -g1 /Users/dilum/dev/repos/JuliaLang/julia/usr/bin/../share/julia/test/runtests.jl Profile`,["CLICOLOR=1", "XPC_FLAGS=0x0", "COMMAND_MODE=unix2003", "PATH=/Users/dilum/.asdf/shims:/Users/dilum/.asdf/bin:~/bin:~/.cargo/bin:/usr/local/sbin:/Applications/Racket v7.9/bin:/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/usr/local/MacGPG2/bin:/opt/X11/bin:/Library/Apple/usr/bin", "PWD=/Users/dilum/dev/repos/JuliaLang/julia", "PYTHON=", "DISPLAY=/private/tmp/com.apple.launchd.A0EIDA2sw6/org.xquartz:0", "XPC_SERVICE_NAME=0", "TERM_PROGRAM=iTerm.app", "SHELL=/bin/bash"  …  "USER=dilum", "JULIA_CUDA_VERBOSE=true", "ITERM_SESSION_ID=w1t0p0:C20FA3E4-4F01-4461-8C4E-18134712344B", "ASDF_DIR=/Users/dilum/.asdf", "TERM=xterm-256color", "HOME=/Users/dilum", "TERM_PROGRAM_VERSION=3.4.8", "JULIA_NUM_THREADS=16", "COLORTERM=truecolor", "OPENBLAS_MAIN_FREE=1"]), ProcessExited(1)) [1]

Stacktrace:
 [1] pipeline_error
   @ ./process.jl:529 [inlined]
 [2] run(::Cmd; wait::Bool)
   @ Base ./process.jl:444
 [3] run
   @ ./process.jl:442 [inlined]
 [4] runtests(tests::Vector{String}; ncores::Int64, exit_on_error::Bool, revise::Bool, seed::Nothing)
   @ Base ./util.jl:570
 [5] top-level scope
   @ none:1
```

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes https://github.com/JuliaLang/julia/issues/42252

@DilumAluthge can your buildkite setup be tested with this before merge?

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by DilumAluthge** [comment]
Yes I'll set up a job to test this PR on Buildkite.

**Comment by DilumAluthge** [comment]
Also, when was threading support added to the profiler? We will need to backport this PR to those releases.

**Comment by IanButterworth** [comment]
The relevant changes are only on master. Profile has worked on multiple threads for a while, but wrt the relevant stuff here, it's just master (1.8...). No need to backport anything

**Comment by DilumAluthge** [comment]
Buildkite test: https://github.com/JuliaLang/julia/pull/42281

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -55,7 +55,7 @@ function init(; n::Union{Nothing,Integer} = nothing, delay::Union{Nothing,Real}
     n_cur = ccall(:jl_profile_maxlen_data, Csize_t, ())
     delay_cur = ccall(:jl_profile_delay_nsec, UInt64, ())/10^9
     if n === nothing && delay === nothing
-        return Int(n_cur), delay_cur
+        return round(Int, n_cur / Threads.nthreads()), delay_cur
```
Comment (DilumAluthge) [inline stdlib/Profile/src/Profile.jl:58]: Should this be `round(Int, n_cur), delay_cur` on Windows, and `round(Int, n_cur / Threads.nthreads()), delay_cur` on non-Windows?

Comment (IanButterworth) [reply stdlib/Profile/src/Profile.jl:58]: Ah, yes, good point. I'll fix that

**Comment by DilumAluthge** [comment]
I'm still seeing a test failure on Buildkite. See https://github.com/JuliaLang/julia/pull/42281

**Comment by IanButterworth** [comment]
This test was failing because now 17 is multiplied by nthreads, so the certain failure this setup relied on was no longer the case with `nthreads > 1`
```
Profile.init(n=17)
@test_logs (:warn, r"The profile data buffer is full") Profile.fetch()
```

So I made the buffer declare itself full if there's not enough space to sample all threads, which I think makes sense

**Comment by DilumAluthge** [comment]
There's one failure left still.

```
Profile                             (3) |         failed at 2021-09-17T01:47:04.886
Test Failed at /cache/build/amdci5-0/julialang/julia-master/julia-e8f1864c1a/share/julia/stdlib/v1.8/Profile/test/runtests.jl:117
  Expression: n_ == 10000000
   Evaluated: 8388608 == 10000000
 
```

https://buildkite.com/julialang/julia-master/builds/3737

**Comment by IanButterworth** [comment]
🤞

**Comment by DilumAluthge** [comment]
Hmmm... still failing 😢 

https://buildkite.com/julialang/julia-master/builds/3750#5f7a271e-6cdd-47ac-8c36-7702d877b312

```
Profile                            (12) |         failed at 2021-09-17T04:23:12.429
Test Failed at /cache/build/amdci7-0/julialang/julia-master/julia-83fe238c3a/share/julia/stdlib/v1.8/Profile/test/runtests.jl:115
  Expression: n_ == 1000000
   Evaluated: 8388608 == 1000000
```

**Comment by DilumAluthge** [comment]
I looked at the changes in https://github.com/IanButterworth/julia/commit/bc16168aeea0f69b6adb56b6f51574f0601f41d5

The conditionals modified in that commit look like this:

```
if 1_000_000 * nthreads > 2^29
    @test n_ * nthreads <= 2^29
else
    @test n_ == 1_000_000
end
```

```
if 1_000_001 * nthreads > 2^29
    @test n_ * nthreads <= 2^29
else
    @test n_ == 1_000_001
end
```

On the `linux32` Buildkite job, the `nthreads` is `16`. So the `1_000_000 * nthreads > 2^29` and ` 1_000_000 * nthreads > 2^29` will both evaluate to `false`, and the `else` will be taken

**Comment by DilumAluthge** [comment]
Hmmm. Won't the conditional still be `false`?

E.g. on `linux32`, `sizeof(Ptr)` is `4`, right? So `1_000_000 * nthreads * sample_size_bytes = 1_000_000 * 16 * 4 = 1_000_000 * 16 * 4 = 64000000`, which is still less than `2^29`.

**Comment by DilumAluthge** [comment]
So, based on https://github.com/JuliaLang/julia/blob/12c0d8423bc91b1a3863ec0dba48c824613c6c11/stdlib/Profile/src/Profile.jl#L92

Should the `1_000_000` on line 113 (https://github.com/JuliaLang/julia/blob/12c0d8423bc91b1a3863ec0dba48c824613c6c11/stdlib/Profile/test/runtests.jl#L113) be `1_000_000` if we are on Windows 32-bit, and `10_000_000` otherwise?

File: stdlib/Profile/test/runtests.jl [hunk 2]
```diff
@@ -106,15 +106,34 @@ end
 
 @testset "setting sample count and delay in init" begin
     n_, delay_ = Profile.init()
-    def_n = Sys.iswindows() && Sys.WORD_SIZE == 32 ? 1_000_000 : 10_000_000
-    @test n_ == def_n
+    n_original = n_
+    nthreads = Sys.iswindows() ? 1 : Threads.nthreads()
+    sample_size_bytes = sizeof(Ptr)
+    if Sys.WORD_SIZE == 32
+        if 1_000_000 * nthreads * sample_size_bytes > 2^29
```
Comment (DilumAluthge) [inline stdlib/Profile/test/runtests.jl:113]:
```suggestion
    def_n = Sys.iswindows() && Sys.WORD_SIZE == 32 ? 1_000_000 : 10_000_000
        if def_n * nthreads * sample_size_bytes > 2^29
```

Comment (DilumAluthge) [inline stdlib/Profile/test/runtests.jl:113]:
```suggestion
    def_n = Sys.iswindows() && Sys.WORD_SIZE == 32 ? 1_000_000 : 10_000_000
        if def_n * nthreads * sample_size_bytes > 2^29
```

**Comment by IanButterworth** [comment]
I've not been giving this full attention, apologies. I think it should be done now. If not, I can revisit over the weekend

**Comment by DilumAluthge** [comment]
No need to apologize at all! I really appreciate you doing all of this work so quickly.

**Comment by DilumAluthge** [comment]
The latest Buildkite run looks good! See https://github.com/JuliaLang/julia/pull/42281 and https://buildkite.com/julialang/julia-master/builds/3759.

**Pull request merged**